{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "A7FsMCVfBXOY"
      },
      "source": [
        "# Final Project of Introduction to Bioinformatics\n",
        "\n",
        "## Comparative Mitochondrial Genomics and Phylogenetic Analysis\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6N11eHGXBXOa"
      },
      "source": [
        "This task focuses on the comparative analysis of mitochondrial genomes from different species, primarily birds, mammals, and insects. The aim is to understand the evolutionary relationships between these species by analyzing and comparing their mitochondrial DNA, which is about 16,000 base pairs in length. You will use advanced computational methods to construct phylogenetic trees and delve into the ecological and anthropological insights that can be gleaned from this data. This project is designed to provide a comprehensive understanding of mitochondrial genomics, its importance in evolutionary biology, and its applications in broader scientific contexts.\n",
        "\n",
        "You will learn:\n",
        "\n",
        "- Techniques for aligning and comparing mitochondrial DNA sequences.\n",
        "- How to construct and interpret phylogenetic trees using advanced computational methods.\n",
        "- The application of mitochondrial genomics in understanding ecological interactions and human evolutionary history.\n",
        "\n",
        "#### Task Roadmap\n",
        "\n",
        "1. **Mitochondrial Genome Comparison**:\n",
        "   - Align mitochondrial DNA sequences from the provided dataset.\n",
        "   - Analyze these sequences to identify similarities and differences across species.\n",
        "\n",
        "2. **Phylogenetic Analysis Using Advanced Methods**:\n",
        "   - Apply Maximum Likelihood (ML) and Bayesian Inference methods, utilizing tools like `ETE Toolkit`, `DendroPy`, `BEAST`, or `PyRate`.\n",
        "   - Compare the trees generated by these methods to understand how different approaches can lead to different interpretations of the data.\n",
        "\n",
        "3. **Cross-Disciplinary Applications (Bonus)**:\n",
        "   - **Ecology**: Examine how mitochondrial DNA analysis can reveal information about species adaptation, migration, and conservation. This involves understanding how genetic variation within and between species can inform ecological strategies and conservation efforts.\n",
        "   - **Anthropology**: Investigate the use of mitochondrial DNA in tracing human evolution and migration patterns. This includes studying the mitochondrial DNA of mammals in your dataset to draw parallels with human evolutionary studies.\n",
        "\n",
        "### Data Sources\n",
        "\n",
        "The mitochondrial DNA data for birds, mammals, and insects will be provided to you. This dataset has been curated to facilitate a comprehensive comparative analysis and is essential for the completion of your phylogenetic studies.\n",
        "\n",
        "### Useful Resources and Material\n",
        "\n",
        "- [Mitochondrial DNA - Wikipedia](https://en.wikipedia.org/wiki/Mitochondrial_DNA): A general introduction to the structure, function, origin, and diversity of mitochondrial DNA, as well as its applications in various fields such as medicine, forensics, and anthropology.\n",
        "- [Mitochondrial DNA Analysis: Introduction, Methods, and Applications](https://bioinfo.cd-genomics.com/mitochondrial-dna-analysis-introduction-methods-and-applications.html): An explanation of the basics of mitochondrial DNA sequencing, bioinformatics analysis, heteroplasmy, and advantages of mitochondrial DNA analysis over nuclear DNA analysis.\n",
        "- [Phylogenetic Tree- Definition, Types, Steps, Methods, Uses - Microbe Notes](https://microbenotes.com/phylogenetic-tree/): A coverage of the concepts and methods of phylogenetic tree construction, including the types of phylogenetic trees, the steps involved in phylogenetic analysis, the main methods of phylogenetic inference, and the applications of phylogenetic trees in various disciplines.\n",
        "- [Phylogenetics - Wikipedia](https://en.wikipedia.org/wiki/Phylogenetics): An overview of the field of phylogenetics, which is the study of the evolutionary history and relationships among or within groups of organisms. It also discusses the data sources, models, algorithms, software, and challenges of phylogenetic analysis.\n",
        "- [ETE Toolkit](http://etetoolkit.org/): A Python library for manipulating, analyzing, and visualizing phylogenetic trees. It supports various formats, methods, and tools for phylogenetic analysis, such as alignment, tree inference, tree comparison, tree annotation, and tree visualization.\n",
        "- [DendroPy](https://dendropy.org/): Another Python library for phylogenetic computing. It provides a comprehensive API for working with phylogenetic data structures, such as trees, characters, and networks. It also offers a rich set of functions for simulation, manipulation, analysis, and annotation of phylogenetic data."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "fYnmE2C9BXOb"
      },
      "source": [
        "### Exploration and Reflection\n",
        "\n",
        "As we proceed with our analysis of mitochondrial DNA for phylogenetic tree construction, it is valuable to contemplate a few questions. These inquiries aim to facilitate a more thorough understanding of the roles and characteristics of mitochondrial DNA in the context of evolutionary biology:\n",
        "\n",
        "1. **Maternal Inheritance and Its Implications**: How does the maternal inheritance of mitochondrial DNA simplify our understanding of evolutionary lineage compared to nuclear DNA, which undergoes recombination? What unique insights can this aspect provide in tracing the evolutionary history of species?\n",
        "\n",
        "2. **Mutation Rate and Evolutionary Insights**: Mitochondrial DNA mutates at a faster rate than nuclear DNA. How does this characteristic make mtDNA a more sensitive tool for detecting recent evolutionary events and relationships among closely related species? Can you think of any specific scenarios or studies where this property of mtDNA has been particularly instrumental?\n",
        "\n",
        "Reflect on these questions as you work through the project, and consider how the properties of mitochondrial DNA enhance its value and applicability in evolutionary biology and beyond. Provide your answer either in this notebook, or in your report (if you had one).\n",
        "\n",
        "<blockquote style=\"font-family:Arial; color:red; font-size:16px; border-left:0px solid red; padding: 10px;\">\n",
        "    <strong>Don't forget to answer these questions!</strong>\n",
        "</blockquote>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "question 1:\n",
        "\n",
        "Maternal inheritance of mitochondrial DNA (mtDNA) offers several simplifications compared to nuclear DNA, which undergoes recombination. Here's how:\n",
        "\n",
        "1- Uniparental Inheritance: Mitochondrial DNA is inherited exclusively from the mother. This uniparental inheritance contrasts with nuclear DNA, which is inherited from both parents and undergoes recombination during sexual reproduction. Because mtDNA is solely passed from mother to offspring, it provides a clearer, more direct line of descent.\n",
        "\n",
        "2- Lack of Recombination: Unlike nuclear DNA, which undergoes recombination during meiosis, mtDNA typically does not recombine. This lack of recombination means that the genetic information in mtDNA remains relatively stable across generations. It simplifies the tracing of maternal lineages over evolutionary time scales since changes in mtDNA are due primarily to mutation rather than recombination.\n",
        "\n",
        "These simplifications offer unique insights into tracing the evolutionary history of species:\n",
        "\n",
        "1- Maternal Lineage Tracing: Because mtDNA is passed exclusively from mother to offspring, it provides a direct line of descent through the maternal lineage. This can be particularly useful for studying population genetics, understanding migration patterns, and reconstructing evolutionary relationships among populations and species.\n",
        "\n",
        "2- Molecular Clock: The relatively stable mutation rate of mtDNA allows researchers to use it as a molecular clock to estimate divergence times between different populations or species. By comparing the differences in mtDNA sequences between individuals or groups and assuming a constant rate of mutation, scientists can infer when populations diverged from a common ancestor.\n",
        "\n",
        "3- Population History and Demography: Patterns of genetic variation in mtDNA can reveal information about population history, such as population size changes, migration events, and demographic fluctuations. Analyzing mtDNA diversity within and among populations helps researchers understand how populations have evolved and adapted to changing environments over time.\n",
        "\n",
        "4- Species Evolution and Phylogenetics: Studying mtDNA sequences from different species provides insights into evolutionary relationships and phylogenetic history. By comparing mtDNA sequences among species, researchers can reconstruct evolutionary trees and understand the processes driving species diversification and adaptation."
      ],
      "metadata": {
        "id": "QTyBVq9jReZS"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "question 2:\n",
        "\n",
        "The faster mutation rate of mitochondrial DNA (mtDNA) compared to nuclear DNA indeed makes it a more sensitive tool for detecting recent evolutionary events and relationships among closely related species. Here's how:\n",
        "\n",
        "Higher Mutation Rate: Mitochondrial DNA accumulates mutations at a faster rate than nuclear DNA due to various factors such as lack of efficient DNA repair mechanisms and exposure to reactive oxygen species within the mitochondria. This faster mutation rate means that changes in mtDNA sequences occur more rapidly over relatively short evolutionary timescales.\n",
        "\n",
        "Recent Evolutionary Events: Because mtDNA evolves more rapidly, it is particularly useful for studying recent evolutionary events, such as population divergence, speciation events, and demographic changes. By examining the differences in mtDNA sequences among individuals or populations, researchers can infer recent evolutionary processes and understand how populations have diverged and adapted in response to environmental changes.\n",
        "\n",
        "Closely Related Species: MtDNA is especially valuable for resolving relationships among closely related species or populations where nuclear DNA sequences may not provide enough genetic variation to distinguish between them. The higher mutation rate of mtDNA increases the likelihood of observing sequence differences that reflect recent evolutionary divergence, allowing researchers to more accurately reconstruct phylogenetic relationships and evolutionary history.\n",
        "\n",
        "Specific scenarios and studies where the faster mutation rate of mtDNA has been instrumental include:\n",
        "\n",
        "Population Genetics: Studies investigating the genetic structure and diversity of populations, particularly in species with rapid population expansions or contractions, often rely on mtDNA due to its sensitivity to recent demographic events.\n",
        "\n",
        "Phylogeography: Research into the historical biogeography and migration patterns of species frequently employs mtDNA to trace the movement of populations and identify refugia during periods of climatic change.\n",
        "\n",
        "Species Delimitation: In taxonomically challenging groups where species boundaries are unclear, mtDNA sequencing can help distinguish between closely related species and identify cryptic species that are morphologically similar but genetically distinct.\n",
        "\n",
        "Forensic Analysis: MtDNA analysis is commonly used in forensic genetics to identify individuals, especially in cases where nuclear DNA may be degraded or contaminated. Its rapid evolution allows for the detection of subtle genetic differences between individuals or closely related family members"
      ],
      "metadata": {
        "id": "MyxvseKHrC_G"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0h6RuwH0BXOb"
      },
      "source": [
        "### Step 0: Installing Necessary Packages"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "P6cnCOQ-BXOc",
        "outputId": "b8ab350c-1aa7-448d-ac78-502babe85e3e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "biopython is NOT installed\n",
            "biopython was successfully installed.\n",
            "pandas (1.5.3) is installed\n",
            "numpy (1.23.5) is installed\n"
          ]
        }
      ],
      "source": [
        "import sys\n",
        "import subprocess\n",
        "import pkg_resources\n",
        "\n",
        "def install(package):\n",
        "    subprocess.check_call([sys.executable, \"-m\", \"pip\", \"install\", package])\n",
        "\n",
        "REQUIRED_PACKAGES = [\n",
        "    'biopython',\n",
        "    'pandas',\n",
        "    'numpy'\n",
        "]\n",
        "\n",
        "for package in REQUIRED_PACKAGES:\n",
        "    try:\n",
        "        dist = pkg_resources.get_distribution(package)\n",
        "        print('{} ({}) is installed'.format(dist.key, dist.version))\n",
        "    except pkg_resources.DistributionNotFound:\n",
        "        print('{} is NOT installed'.format(package))\n",
        "        install(package)\n",
        "        print('{} was successfully installed.'.format(package))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "id": "oo7zEZQVBXOe"
      },
      "outputs": [],
      "source": [
        "# Import necessary libraries.\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "from Bio import Entrez\n",
        "import re\n",
        "import time\n",
        "from tqdm import tqdm"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xK55L8ujBXOe"
      },
      "source": [
        "### Step 1: Dataset Expansion\n",
        "\n",
        "Our first task is to augment our dataset with additional species. This involves engaging with the NCBI database to retrieve mitochondrial DNA sequences.\n",
        "\n",
        "#### Instructions:\n",
        "\n",
        "- **Species Selection**: Identify and choose 10 additional species to include in your dataset. Aim for a diverse selection to enrich your phylogenetic analysis.\n",
        "\n",
        "- **Querying NCBI Database**: Use the NCBI database to locate mitochondrial DNA sequences for your chosen species. While you can manually search on the [NCBI website](https://www.ncbi.nlm.nih.gov/), consider automating this process through their API for a more efficient approach.\n",
        "    - **Example Query**: As a starting point, you might use a query like `\"mitochondrion[Filter] AND (your_species_name[Organism])` to find specific mtDNA sequences. Adjust the query parameters according to your species selection.\n",
        "    - **Documentation**: Familiarize yourself with the [NCBI API documentation](https://www.ncbi.nlm.nih.gov/books/NBK25497/) for detailed guidance on constructing queries.\n",
        "\n",
        "- **Using NCBI Website**: You are welcome to use the NCBI website for this task. If you do so, document each step of your process clearly in your task report. This should include the species names, search terms used, and how you determined the relevant sequences to include.\n",
        "\n",
        "- **Bonus Opportunity**: Implementing an automated, methodological approach using the NCBI API and relevant Python packages to add all 10 records in your dataset will earn you a 50% bonus for this section. Your method should be structured and replicable, demonstrating a systematic approach to data collection.\n",
        "\n",
        "Remember, the goal is to methodically expand your dataset with relevant mtDNA sequences, paving the way for insightful phylogenetic analysis."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 788
        },
        "id": "XZ52vZ85BXOe",
        "outputId": "bdb5b43c-ae6b-4e41-81d5-6789b4865e2f"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "(23, 6)\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "    taxo_id                    specie      blast_name  genbank_common_name  \\\n",
              "0      8945     Eudynamys scolopaceus           birds           Asian koel   \n",
              "1      7460            Apis mellifera            Bees            honey bee   \n",
              "2     36300         Pelecanus crispus           birds    Dalmatian pelican   \n",
              "3     10116         Rattus norvegicus         Rodents           Norway rat   \n",
              "4      9031             Gallus gallus           birds        Gallus gallus   \n",
              "5     59463          Myotis lucifugus            bats     little brown bat   \n",
              "6      7070       Tribolium castaneum         Insects  Tribolium castaneum   \n",
              "7      8845           Anser cygnoides           birds           Swan goose   \n",
              "8      9940                Ovis aries           sheep  even-toed ungulates   \n",
              "9     56313                 Tyto alba           birds      common barn owl   \n",
              "10     7227   Drosophila melanogaster           flies            fruit fly   \n",
              "11     8822         Apteryx australis           birds  Southern brown kiwi   \n",
              "12    92683       Spheniscus demersus           birds      jackass penguin   \n",
              "13    85066     Corvus brachyrhynchos           birds        American crow   \n",
              "14    10090              Mus musculus         rodents          house mouse   \n",
              "15     8790  Dromaius novaehollandiae           birds                  emu   \n",
              "16     8839        Anas platyrhynchos           birds              mallard   \n",
              "17   132908         Pteropus vampyrus            bats     large flying fox   \n",
              "18     8801          Struthio camelus           birds      African ostrich   \n",
              "19   193005         Aquila nipalensis  hawks & eagles         steppe eagle   \n",
              "20   219595             Cygnus cygnus           birds       common whooper   \n",
              "21    59729       Taeniopygia guttata           birds          zebra finch   \n",
              "22     9233      Aptenodytes forsteri           birds      emperor penguin   \n",
              "\n",
              "   accession_number                                              mtDNA  \n",
              "0         NC_060520  https://www.ncbi.nlm.nih.gov/nucleotide/NC_060...  \n",
              "1         NC_051932  https://www.ncbi.nlm.nih.gov/nucleotide/NC_051...  \n",
              "2          OR620163    https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1  \n",
              "3         NC_001665     https://www.ncbi.nlm.nih.gov/nuccore/NC_001665  \n",
              "4         NC_053523   https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1  \n",
              "5         NC_029849   https://www.ncbi.nlm.nih.gov/nuccore/NC_029849.1  \n",
              "6         NC_003081     https://www.ncbi.nlm.nih.gov/nuccore/NC_003081  \n",
              "7         NC_023832  https://www.ncbi.nlm.nih.gov/nucleotide/NC_023...  \n",
              "8         NC_001941     https://www.ncbi.nlm.nih.gov/nuccore/NC_001941  \n",
              "9          MZ318036    https://www.ncbi.nlm.nih.gov/nuccore/MZ318036.1  \n",
              "10        NC_024511  https://www.ncbi.nlm.nih.gov/nucleotide/NC_024...  \n",
              "11         MN356385    https://www.ncbi.nlm.nih.gov/nuccore/MN356385.1  \n",
              "12        NC_022817  https://www.ncbi.nlm.nih.gov/nucleotide/NC_022...  \n",
              "13        NC_026461  https://www.ncbi.nlm.nih.gov/nucleotide/NC_026...  \n",
              "14        NC_005089     https://www.ncbi.nlm.nih.gov/nuccore/NC_005089  \n",
              "15        NC_002784     https://www.ncbi.nlm.nih.gov/nuccore/NC_002784  \n",
              "16        NC_009684     https://www.ncbi.nlm.nih.gov/nuccore/NC_009684  \n",
              "17        NC_026542   https://www.ncbi.nlm.nih.gov/nuccore/NC_026542.1  \n",
              "18        NC_002785     https://www.ncbi.nlm.nih.gov/nuccore/NC_002785  \n",
              "19        NC_045042   https://www.ncbi.nlm.nih.gov/nuccore/NC_045042.1  \n",
              "20        NC_027095  https://www.ncbi.nlm.nih.gov/nucleotide/NC_027...  \n",
              "21        NC_007897  https://www.ncbi.nlm.nih.gov/nucleotide/NC_007...  \n",
              "22        NC_027938  https://www.ncbi.nlm.nih.gov/nucleotide/NC_027...  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-ea71a954-1eed-4ff0-aa05-bba08a985196\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>taxo_id</th>\n",
              "      <th>specie</th>\n",
              "      <th>blast_name</th>\n",
              "      <th>genbank_common_name</th>\n",
              "      <th>accession_number</th>\n",
              "      <th>mtDNA</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>8945</td>\n",
              "      <td>Eudynamys scolopaceus</td>\n",
              "      <td>birds</td>\n",
              "      <td>Asian koel</td>\n",
              "      <td>NC_060520</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_060...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>7460</td>\n",
              "      <td>Apis mellifera</td>\n",
              "      <td>Bees</td>\n",
              "      <td>honey bee</td>\n",
              "      <td>NC_051932</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_051...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>36300</td>\n",
              "      <td>Pelecanus crispus</td>\n",
              "      <td>birds</td>\n",
              "      <td>Dalmatian pelican</td>\n",
              "      <td>OR620163</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>10116</td>\n",
              "      <td>Rattus norvegicus</td>\n",
              "      <td>Rodents</td>\n",
              "      <td>Norway rat</td>\n",
              "      <td>NC_001665</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_001665</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>9031</td>\n",
              "      <td>Gallus gallus</td>\n",
              "      <td>birds</td>\n",
              "      <td>Gallus gallus</td>\n",
              "      <td>NC_053523</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>59463</td>\n",
              "      <td>Myotis lucifugus</td>\n",
              "      <td>bats</td>\n",
              "      <td>little brown bat</td>\n",
              "      <td>NC_029849</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_029849.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>7070</td>\n",
              "      <td>Tribolium castaneum</td>\n",
              "      <td>Insects</td>\n",
              "      <td>Tribolium castaneum</td>\n",
              "      <td>NC_003081</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_003081</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>8845</td>\n",
              "      <td>Anser cygnoides</td>\n",
              "      <td>birds</td>\n",
              "      <td>Swan goose</td>\n",
              "      <td>NC_023832</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_023...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>9940</td>\n",
              "      <td>Ovis aries</td>\n",
              "      <td>sheep</td>\n",
              "      <td>even-toed ungulates</td>\n",
              "      <td>NC_001941</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_001941</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>56313</td>\n",
              "      <td>Tyto alba</td>\n",
              "      <td>birds</td>\n",
              "      <td>common barn owl</td>\n",
              "      <td>MZ318036</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/MZ318036.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>7227</td>\n",
              "      <td>Drosophila melanogaster</td>\n",
              "      <td>flies</td>\n",
              "      <td>fruit fly</td>\n",
              "      <td>NC_024511</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_024...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>8822</td>\n",
              "      <td>Apteryx australis</td>\n",
              "      <td>birds</td>\n",
              "      <td>Southern brown kiwi</td>\n",
              "      <td>MN356385</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/MN356385.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>92683</td>\n",
              "      <td>Spheniscus demersus</td>\n",
              "      <td>birds</td>\n",
              "      <td>jackass penguin</td>\n",
              "      <td>NC_022817</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_022...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>85066</td>\n",
              "      <td>Corvus brachyrhynchos</td>\n",
              "      <td>birds</td>\n",
              "      <td>American crow</td>\n",
              "      <td>NC_026461</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_026...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>10090</td>\n",
              "      <td>Mus musculus</td>\n",
              "      <td>rodents</td>\n",
              "      <td>house mouse</td>\n",
              "      <td>NC_005089</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_005089</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>8790</td>\n",
              "      <td>Dromaius novaehollandiae</td>\n",
              "      <td>birds</td>\n",
              "      <td>emu</td>\n",
              "      <td>NC_002784</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_002784</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>8839</td>\n",
              "      <td>Anas platyrhynchos</td>\n",
              "      <td>birds</td>\n",
              "      <td>mallard</td>\n",
              "      <td>NC_009684</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_009684</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>132908</td>\n",
              "      <td>Pteropus vampyrus</td>\n",
              "      <td>bats</td>\n",
              "      <td>large flying fox</td>\n",
              "      <td>NC_026542</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_026542.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>8801</td>\n",
              "      <td>Struthio camelus</td>\n",
              "      <td>birds</td>\n",
              "      <td>African ostrich</td>\n",
              "      <td>NC_002785</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_002785</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>193005</td>\n",
              "      <td>Aquila nipalensis</td>\n",
              "      <td>hawks &amp; eagles</td>\n",
              "      <td>steppe eagle</td>\n",
              "      <td>NC_045042</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_045042.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>219595</td>\n",
              "      <td>Cygnus cygnus</td>\n",
              "      <td>birds</td>\n",
              "      <td>common whooper</td>\n",
              "      <td>NC_027095</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_027...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>59729</td>\n",
              "      <td>Taeniopygia guttata</td>\n",
              "      <td>birds</td>\n",
              "      <td>zebra finch</td>\n",
              "      <td>NC_007897</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_007...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>9233</td>\n",
              "      <td>Aptenodytes forsteri</td>\n",
              "      <td>birds</td>\n",
              "      <td>emperor penguin</td>\n",
              "      <td>NC_027938</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_027...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-ea71a954-1eed-4ff0-aa05-bba08a985196')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-ea71a954-1eed-4ff0-aa05-bba08a985196 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-ea71a954-1eed-4ff0-aa05-bba08a985196');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-1b2d5b57-095d-4069-b4ac-1a489c2f8d73\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-1b2d5b57-095d-4069-b4ac-1a489c2f8d73')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-1b2d5b57-095d-4069-b4ac-1a489c2f8d73 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ]
          },
          "metadata": {},
          "execution_count": 3
        }
      ],
      "source": [
        "dataset = pd.read_csv('./dataset/species.csv')\n",
        "\n",
        "print(dataset.shape)\n",
        "dataset"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import time\n",
        "import re\n",
        "import pandas as pd\n",
        "from Bio import Entrez\n",
        "\n",
        "# Set your email address and API key for NCBI Entrez\n",
        "Entrez.email = 'mahdi139460@gmail.com'\n",
        "Entrez.api_key = '833b42e2dce5b722fa613375031f546b3e08'\n",
        "\n",
        "def search_mitochondrial_dna(species_name):\n",
        "    \"\"\"\n",
        "    Search NCBI database for mitochondrial DNA sequences of a given species.\n",
        "    Returns the first NCBI ID if found, otherwise returns None.\n",
        "    \"\"\"\n",
        "    query = f\"mitochondrion[Filter] AND ({species_name}[Organism])\"\n",
        "    handle = Entrez.esearch(db=\"nuccore\", term=query, retmax=1)\n",
        "    record = Entrez.read(handle)\n",
        "    return record[\"IdList\"][0] if record[\"Count\"] != \"0\" else None\n",
        "\n",
        "def fetch_data(db_name, ncbi_id):\n",
        "    \"\"\"\n",
        "    Fetch data from the specified NCBI database using the given NCBI ID.\n",
        "    \"\"\"\n",
        "    handle = Entrez.efetch(db=db_name, id=ncbi_id, retmode=\"xml\")\n",
        "    return Entrez.read(handle)[0]"
      ],
      "metadata": {
        "id": "HgtWqPqU3CTT"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "id": "LDp7K6R8BXOf",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "303aa466-759e-4444-f92f-f997fe1ded38"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mitochondrial DNA sequences retrieval completed!\n"
          ]
        }
      ],
      "source": [
        "# import pandas as pd\n",
        "# import requests\n",
        "\n",
        "# TODO: Add 10 more species to your dataset\n",
        "# for species in additional_species:\n",
        "    # Fetch mtDNA and add to the dataset\n",
        "\n",
        "\n",
        "species_list = [\n",
        "    \"Panthera leo\",\n",
        "    \"Homo sapiens\",\n",
        "    \"Varanus komodoensis\",\n",
        "    \"Felis catus\",\n",
        "    \"Gallus gallus\",\n",
        "    \"Mus musculus\",\n",
        "    \"Oryctolagus cuniculus\",\n",
        "    \"Rattus norvegicus\",\n",
        "    \"Drosophila melanogaster\",\n",
        "    \"Loxodonta africana\"\n",
        "]\n",
        "\n",
        "species_blast_name = {\n",
        "    \"Panthera leo\": 'carnivores',\n",
        "    \"Homo sapiens\": 'primates',\n",
        "    \"Varanus komodoensis\": 'lizards & snakes',\n",
        "    \"Felis catus\": 'carnivores',\n",
        "    \"Gallus gallus\": 'birds',\n",
        "    \"Mus musculus\": 'rodents',\n",
        "    \"Oryctolagus cuniculus\": 'rabbits & hares',\n",
        "    \"Rattus norvegicus\": 'rodents',\n",
        "    \"Drosophila melanogaster\": 'flies',\n",
        "    \"Loxodonta africana\": 'placentals',\n",
        "}\n",
        "\n",
        "dfs = []\n",
        "\n",
        "for species in species_list:\n",
        "    ncbi_id = search_mitochondrial_dna(species)\n",
        "    if ncbi_id:\n",
        "        try:\n",
        "            species_data = fetch_data('nuccore', ncbi_id)\n",
        "\n",
        "            accession_number = species_data['GBSeq_accession-version']\n",
        "\n",
        "            taxon_id_match = re.search(r\"taxon:(\\d+)\", str(species_data))\n",
        "            taxon_id = taxon_id_match.group(1) if taxon_id_match else None\n",
        "\n",
        "            if taxon_id:\n",
        "                taxonomy_data = fetch_data('taxonomy', taxon_id)\n",
        "                genbank_common_name = taxonomy_data['OtherNames']['GenbankCommonName']\n",
        "            else:\n",
        "                genbank_common_name = None\n",
        "\n",
        "            df_row = pd.DataFrame({\n",
        "                'taxo_id': [taxon_id],\n",
        "                'specie': [species],\n",
        "                'blast_name': [species_blast_name[species]],\n",
        "                'genbank_common_name': [genbank_common_name],\n",
        "                'accession_number': [accession_number.split('.')[0]],\n",
        "                'mtDNA': [f'https://www.ncbi.nlm.nih.gov/nuccore/{accession_number}']\n",
        "            })\n",
        "            dfs.append(df_row)\n",
        "        except Exception as e:\n",
        "            print(f\"Error processing {species}: {e}\")\n",
        "    else:\n",
        "        print(f\"No mitochondrial DNA sequence found for {species}\")\n",
        "\n",
        "result_df = pd.concat(dfs, ignore_index=True)\n",
        "print(\"Mitochondrial DNA sequences retrieval completed!\")\n",
        "\n",
        "\n",
        "expanded_dataset = pd.concat([dataset, result_df], ignore_index=True)\n",
        "\n",
        "# TODO: Save the expanded dataset\n",
        "expanded_dataset.to_csv('./dataset/expanded_dataset.csv', index=False)\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "expanded_dataset"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "ePiEW5Tm6Fau",
        "outputId": "686b01fc-8d4c-45b5-f708-2ed542241585"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "   taxo_id                    specie        blast_name  \\\n",
              "0     8945     Eudynamys scolopaceus             birds   \n",
              "1     7460            Apis mellifera              Bees   \n",
              "2    36300         Pelecanus crispus             birds   \n",
              "3    10116         Rattus norvegicus           Rodents   \n",
              "4     9031             Gallus gallus             birds   \n",
              "5    59463          Myotis lucifugus              bats   \n",
              "6     7070       Tribolium castaneum           Insects   \n",
              "7     8845           Anser cygnoides             birds   \n",
              "8     9940                Ovis aries             sheep   \n",
              "9    56313                 Tyto alba             birds   \n",
              "10    7227   Drosophila melanogaster             flies   \n",
              "11    8822         Apteryx australis             birds   \n",
              "12   92683       Spheniscus demersus             birds   \n",
              "13   85066     Corvus brachyrhynchos             birds   \n",
              "14   10090              Mus musculus           rodents   \n",
              "15    8790  Dromaius novaehollandiae             birds   \n",
              "16    8839        Anas platyrhynchos             birds   \n",
              "17  132908         Pteropus vampyrus              bats   \n",
              "18    8801          Struthio camelus             birds   \n",
              "19  193005         Aquila nipalensis    hawks & eagles   \n",
              "20  219595             Cygnus cygnus             birds   \n",
              "21   59729       Taeniopygia guttata             birds   \n",
              "22    9233      Aptenodytes forsteri             birds   \n",
              "23    9689              Panthera leo        carnivores   \n",
              "24    9606              Homo sapiens          primates   \n",
              "25   61221       Varanus komodoensis  lizards & snakes   \n",
              "26    9685               Felis catus        carnivores   \n",
              "27    9031             Gallus gallus             birds   \n",
              "28   10090              Mus musculus           rodents   \n",
              "29    9986     Oryctolagus cuniculus   rabbits & hares   \n",
              "30   10116         Rattus norvegicus           rodents   \n",
              "31    7227   Drosophila melanogaster             flies   \n",
              "32    9785        Loxodonta africana        placentals   \n",
              "\n",
              "         genbank_common_name accession_number  \\\n",
              "0                 Asian koel        NC_060520   \n",
              "1                  honey bee        NC_051932   \n",
              "2          Dalmatian pelican         OR620163   \n",
              "3                 Norway rat        NC_001665   \n",
              "4              Gallus gallus        NC_053523   \n",
              "5           little brown bat        NC_029849   \n",
              "6        Tribolium castaneum        NC_003081   \n",
              "7                 Swan goose        NC_023832   \n",
              "8        even-toed ungulates        NC_001941   \n",
              "9            common barn owl         MZ318036   \n",
              "10                 fruit fly        NC_024511   \n",
              "11       Southern brown kiwi         MN356385   \n",
              "12           jackass penguin        NC_022817   \n",
              "13             American crow        NC_026461   \n",
              "14               house mouse        NC_005089   \n",
              "15                       emu        NC_002784   \n",
              "16                   mallard        NC_009684   \n",
              "17          large flying fox        NC_026542   \n",
              "18           African ostrich        NC_002785   \n",
              "19              steppe eagle        NC_045042   \n",
              "20            common whooper        NC_027095   \n",
              "21               zebra finch        NC_007897   \n",
              "22           emperor penguin        NC_027938   \n",
              "23                      lion        NC_028302   \n",
              "24                     human         PP153372   \n",
              "25             Komodo dragon         AB080276   \n",
              "26              domestic cat         PP070534   \n",
              "27                   chicken         CM068470   \n",
              "28               house mouse         KY986857   \n",
              "29                    rabbit         OR210205   \n",
              "30                Norway rat         KY986806   \n",
              "31                 fruit fly         OR917921   \n",
              "32  African savanna elephant         OR344290   \n",
              "\n",
              "                                                mtDNA  \n",
              "0   https://www.ncbi.nlm.nih.gov/nucleotide/NC_060...  \n",
              "1   https://www.ncbi.nlm.nih.gov/nucleotide/NC_051...  \n",
              "2     https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1  \n",
              "3      https://www.ncbi.nlm.nih.gov/nuccore/NC_001665  \n",
              "4    https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1  \n",
              "5    https://www.ncbi.nlm.nih.gov/nuccore/NC_029849.1  \n",
              "6      https://www.ncbi.nlm.nih.gov/nuccore/NC_003081  \n",
              "7   https://www.ncbi.nlm.nih.gov/nucleotide/NC_023...  \n",
              "8      https://www.ncbi.nlm.nih.gov/nuccore/NC_001941  \n",
              "9     https://www.ncbi.nlm.nih.gov/nuccore/MZ318036.1  \n",
              "10  https://www.ncbi.nlm.nih.gov/nucleotide/NC_024...  \n",
              "11    https://www.ncbi.nlm.nih.gov/nuccore/MN356385.1  \n",
              "12  https://www.ncbi.nlm.nih.gov/nucleotide/NC_022...  \n",
              "13  https://www.ncbi.nlm.nih.gov/nucleotide/NC_026...  \n",
              "14     https://www.ncbi.nlm.nih.gov/nuccore/NC_005089  \n",
              "15     https://www.ncbi.nlm.nih.gov/nuccore/NC_002784  \n",
              "16     https://www.ncbi.nlm.nih.gov/nuccore/NC_009684  \n",
              "17   https://www.ncbi.nlm.nih.gov/nuccore/NC_026542.1  \n",
              "18     https://www.ncbi.nlm.nih.gov/nuccore/NC_002785  \n",
              "19   https://www.ncbi.nlm.nih.gov/nuccore/NC_045042.1  \n",
              "20  https://www.ncbi.nlm.nih.gov/nucleotide/NC_027...  \n",
              "21  https://www.ncbi.nlm.nih.gov/nucleotide/NC_007...  \n",
              "22  https://www.ncbi.nlm.nih.gov/nucleotide/NC_027...  \n",
              "23   https://www.ncbi.nlm.nih.gov/nuccore/NC_028302.1  \n",
              "24    https://www.ncbi.nlm.nih.gov/nuccore/PP153372.1  \n",
              "25    https://www.ncbi.nlm.nih.gov/nuccore/AB080276.1  \n",
              "26    https://www.ncbi.nlm.nih.gov/nuccore/PP070534.1  \n",
              "27    https://www.ncbi.nlm.nih.gov/nuccore/CM068470.1  \n",
              "28    https://www.ncbi.nlm.nih.gov/nuccore/KY986857.1  \n",
              "29    https://www.ncbi.nlm.nih.gov/nuccore/OR210205.1  \n",
              "30    https://www.ncbi.nlm.nih.gov/nuccore/KY986806.1  \n",
              "31    https://www.ncbi.nlm.nih.gov/nuccore/OR917921.1  \n",
              "32    https://www.ncbi.nlm.nih.gov/nuccore/OR344290.1  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-27fa898e-d09d-41dd-9b97-dccbc3631367\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>taxo_id</th>\n",
              "      <th>specie</th>\n",
              "      <th>blast_name</th>\n",
              "      <th>genbank_common_name</th>\n",
              "      <th>accession_number</th>\n",
              "      <th>mtDNA</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>8945</td>\n",
              "      <td>Eudynamys scolopaceus</td>\n",
              "      <td>birds</td>\n",
              "      <td>Asian koel</td>\n",
              "      <td>NC_060520</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_060...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>7460</td>\n",
              "      <td>Apis mellifera</td>\n",
              "      <td>Bees</td>\n",
              "      <td>honey bee</td>\n",
              "      <td>NC_051932</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_051...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>36300</td>\n",
              "      <td>Pelecanus crispus</td>\n",
              "      <td>birds</td>\n",
              "      <td>Dalmatian pelican</td>\n",
              "      <td>OR620163</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>10116</td>\n",
              "      <td>Rattus norvegicus</td>\n",
              "      <td>Rodents</td>\n",
              "      <td>Norway rat</td>\n",
              "      <td>NC_001665</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_001665</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>9031</td>\n",
              "      <td>Gallus gallus</td>\n",
              "      <td>birds</td>\n",
              "      <td>Gallus gallus</td>\n",
              "      <td>NC_053523</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>59463</td>\n",
              "      <td>Myotis lucifugus</td>\n",
              "      <td>bats</td>\n",
              "      <td>little brown bat</td>\n",
              "      <td>NC_029849</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_029849.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>7070</td>\n",
              "      <td>Tribolium castaneum</td>\n",
              "      <td>Insects</td>\n",
              "      <td>Tribolium castaneum</td>\n",
              "      <td>NC_003081</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_003081</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>8845</td>\n",
              "      <td>Anser cygnoides</td>\n",
              "      <td>birds</td>\n",
              "      <td>Swan goose</td>\n",
              "      <td>NC_023832</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_023...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>9940</td>\n",
              "      <td>Ovis aries</td>\n",
              "      <td>sheep</td>\n",
              "      <td>even-toed ungulates</td>\n",
              "      <td>NC_001941</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_001941</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>56313</td>\n",
              "      <td>Tyto alba</td>\n",
              "      <td>birds</td>\n",
              "      <td>common barn owl</td>\n",
              "      <td>MZ318036</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/MZ318036.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>7227</td>\n",
              "      <td>Drosophila melanogaster</td>\n",
              "      <td>flies</td>\n",
              "      <td>fruit fly</td>\n",
              "      <td>NC_024511</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_024...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>8822</td>\n",
              "      <td>Apteryx australis</td>\n",
              "      <td>birds</td>\n",
              "      <td>Southern brown kiwi</td>\n",
              "      <td>MN356385</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/MN356385.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>92683</td>\n",
              "      <td>Spheniscus demersus</td>\n",
              "      <td>birds</td>\n",
              "      <td>jackass penguin</td>\n",
              "      <td>NC_022817</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_022...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>85066</td>\n",
              "      <td>Corvus brachyrhynchos</td>\n",
              "      <td>birds</td>\n",
              "      <td>American crow</td>\n",
              "      <td>NC_026461</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_026...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>10090</td>\n",
              "      <td>Mus musculus</td>\n",
              "      <td>rodents</td>\n",
              "      <td>house mouse</td>\n",
              "      <td>NC_005089</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_005089</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>8790</td>\n",
              "      <td>Dromaius novaehollandiae</td>\n",
              "      <td>birds</td>\n",
              "      <td>emu</td>\n",
              "      <td>NC_002784</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_002784</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>8839</td>\n",
              "      <td>Anas platyrhynchos</td>\n",
              "      <td>birds</td>\n",
              "      <td>mallard</td>\n",
              "      <td>NC_009684</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_009684</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>132908</td>\n",
              "      <td>Pteropus vampyrus</td>\n",
              "      <td>bats</td>\n",
              "      <td>large flying fox</td>\n",
              "      <td>NC_026542</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_026542.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>8801</td>\n",
              "      <td>Struthio camelus</td>\n",
              "      <td>birds</td>\n",
              "      <td>African ostrich</td>\n",
              "      <td>NC_002785</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_002785</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>193005</td>\n",
              "      <td>Aquila nipalensis</td>\n",
              "      <td>hawks &amp; eagles</td>\n",
              "      <td>steppe eagle</td>\n",
              "      <td>NC_045042</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_045042.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>219595</td>\n",
              "      <td>Cygnus cygnus</td>\n",
              "      <td>birds</td>\n",
              "      <td>common whooper</td>\n",
              "      <td>NC_027095</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_027...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>59729</td>\n",
              "      <td>Taeniopygia guttata</td>\n",
              "      <td>birds</td>\n",
              "      <td>zebra finch</td>\n",
              "      <td>NC_007897</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_007...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>9233</td>\n",
              "      <td>Aptenodytes forsteri</td>\n",
              "      <td>birds</td>\n",
              "      <td>emperor penguin</td>\n",
              "      <td>NC_027938</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_027...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>9689</td>\n",
              "      <td>Panthera leo</td>\n",
              "      <td>carnivores</td>\n",
              "      <td>lion</td>\n",
              "      <td>NC_028302</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_028302.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>9606</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>primates</td>\n",
              "      <td>human</td>\n",
              "      <td>PP153372</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/PP153372.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>61221</td>\n",
              "      <td>Varanus komodoensis</td>\n",
              "      <td>lizards &amp; snakes</td>\n",
              "      <td>Komodo dragon</td>\n",
              "      <td>AB080276</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/AB080276.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>9685</td>\n",
              "      <td>Felis catus</td>\n",
              "      <td>carnivores</td>\n",
              "      <td>domestic cat</td>\n",
              "      <td>PP070534</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/PP070534.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>9031</td>\n",
              "      <td>Gallus gallus</td>\n",
              "      <td>birds</td>\n",
              "      <td>chicken</td>\n",
              "      <td>CM068470</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/CM068470.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>10090</td>\n",
              "      <td>Mus musculus</td>\n",
              "      <td>rodents</td>\n",
              "      <td>house mouse</td>\n",
              "      <td>KY986857</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/KY986857.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>9986</td>\n",
              "      <td>Oryctolagus cuniculus</td>\n",
              "      <td>rabbits &amp; hares</td>\n",
              "      <td>rabbit</td>\n",
              "      <td>OR210205</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/OR210205.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>10116</td>\n",
              "      <td>Rattus norvegicus</td>\n",
              "      <td>rodents</td>\n",
              "      <td>Norway rat</td>\n",
              "      <td>KY986806</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/KY986806.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31</th>\n",
              "      <td>7227</td>\n",
              "      <td>Drosophila melanogaster</td>\n",
              "      <td>flies</td>\n",
              "      <td>fruit fly</td>\n",
              "      <td>OR917921</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/OR917921.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td>9785</td>\n",
              "      <td>Loxodonta africana</td>\n",
              "      <td>placentals</td>\n",
              "      <td>African savanna elephant</td>\n",
              "      <td>OR344290</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/OR344290.1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-27fa898e-d09d-41dd-9b97-dccbc3631367')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-27fa898e-d09d-41dd-9b97-dccbc3631367 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-27fa898e-d09d-41dd-9b97-dccbc3631367');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-04cb4ab5-1188-468c-a484-ebd6dc8e6fcb\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-04cb4ab5-1188-468c-a484-ebd6dc8e6fcb')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-04cb4ab5-1188-468c-a484-ebd6dc8e6fcb button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ]
          },
          "metadata": {},
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nyFS0q0GBXOf"
      },
      "source": [
        "#### Checking Data Consistency\n",
        "\n",
        "The code block is designed to check the accuracy of a biological dataset. It examines each entry in a CSV file, focusing on taxonomy IDs, species names, GenBank accession numbers, and mitochondrial DNA links. It uses the NCBI's Entrez system to ensure taxonomy IDs correspond to the correct species and confirms the mitochondrial DNA links are accurate. The script also checks GenBank accession numbers against the provided links. This method is useful for maintaining the accuracy of current data and **might help in adding new entries to the database.**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "id": "PsIwipxTBXOf"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import requests\n",
        "from Bio import Entrez\n",
        "import os\n",
        "from io import BytesIO\n",
        "from requests.adapters import HTTPAdapter\n",
        "from urllib3.util.retry import Retry\n",
        "\n",
        "# Set your email and API key for NCBI\n",
        "# entrez_email = os.getenv('mahdi139460@gmail.com')\n",
        "# entrez_key = os.getenv('833b42e2dce5b722fa613375031f546b3e08')\n",
        "Entrez.email = 'mahdi139460@gmail.com'\n",
        "Entrez.api_key = '833b42e2dce5b722fa613375031f546b3e08'\n",
        "\n",
        "\n",
        "def fetch_entrez_record(db, id, rettype, retmode):\n",
        "    \"\"\"Fetch record from NCBI Entrez with retries.\"\"\"\n",
        "    url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db={db}&id={id}&rettype={rettype}&retmode={retmode}\"\n",
        "    session = requests.Session()\n",
        "    retries = Retry(total=5, backoff_factor=1, status_forcelist=[502, 503, 504])\n",
        "    session.mount('https://', HTTPAdapter(max_retries=retries))\n",
        "\n",
        "    try:\n",
        "        response = session.get(url)\n",
        "        response.raise_for_status()\n",
        "        if db == \"nucleotide\":\n",
        "            return response.text\n",
        "        return response.content\n",
        "    except requests.exceptions.HTTPError as err:\n",
        "        print(f\"HTTP error: {err}\")\n",
        "    except requests.exceptions.ConnectionError as err:\n",
        "        print(f\"Connection error: {err}\")\n",
        "    return None\n",
        "\n",
        "def verify_taxonomy_id(taxo_id, species_name):\n",
        "    \"\"\"Verify taxonomy ID against species name.\"\"\"\n",
        "    xml_data = fetch_entrez_record(\"taxonomy\", taxo_id, \"xml\", \"xml\")\n",
        "    if xml_data:\n",
        "        # Convert bytes data to a binary file-like object\n",
        "        xml_data_io = BytesIO(xml_data)\n",
        "        records = Entrez.read(xml_data_io)\n",
        "        return records[0]['ScientificName'].lower() == species_name.lower()\n",
        "    return False\n",
        "\n",
        "def verify_mitochondrial_dna(accession_number):\n",
        "    gb_data = fetch_entrez_record(\"nucleotide\", accession_number, \"gb\", \"text\")\n",
        "    return \"mitochondrion\" in gb_data.lower() if gb_data else False\n",
        "\n",
        "def extract_accession_from_link(link):\n",
        "    return link.split('/')[-1].split('.')[0]\n",
        "\n",
        "def check_dataset_consistency(file_path):\n",
        "    species_df = pd.read_csv(file_path)\n",
        "\n",
        "    for index, row in species_df.iterrows():\n",
        "        taxonomy_id = str(row['taxo_id'])\n",
        "        species_name = row['specie']\n",
        "        accession_number = row['accession_number']\n",
        "        mtDNA_link = row['mtDNA']\n",
        "        extracted_accession = extract_accession_from_link(mtDNA_link)\n",
        "\n",
        "        taxonomy_check = verify_taxonomy_id(taxonomy_id, species_name)\n",
        "        accession_match = (accession_number == extracted_accession)\n",
        "        mitochondrial_check = verify_mitochondrial_dna(accession_number)\n",
        "\n",
        "        print(f\"Row {index}: Taxonomy Check: {taxonomy_check}, Accession Match: {accession_match}, Mitochondrial Check: {mitochondrial_check}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 74,
      "metadata": {
        "id": "b5YS3S8iBXOg",
        "outputId": "ea2f904e-87fd-4ccf-f928-2848ec037249",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Row 0: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 1: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 2: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 3: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 4: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 5: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 6: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 7: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 8: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 9: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 10: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 11: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 12: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "HTTP error: 400 Client Error: Bad Request for url: https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=taxonomy&id=85066&rettype=xml&retmode=xml\n",
            "Row 13: Taxonomy Check: False, Accession Match: True, Mitochondrial Check: True\n",
            "Row 14: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 15: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 16: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 17: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 18: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 19: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 20: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 21: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 22: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 23: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 24: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 25: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 26: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 27: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 28: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 29: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 30: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 31: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 32: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n"
          ]
        }
      ],
      "source": [
        "check_dataset_consistency('./dataset/expanded_dataset.csv')"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "for index, row in expanded_dataset.iterrows():\n",
        "  print(index, row['mtDNA'])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OXvwDLNN3tXV",
        "outputId": "3beac6ee-4aa2-4edd-9fcb-a84ca61bf89a"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "0 https://www.ncbi.nlm.nih.gov/nucleotide/NC_060520.1\n",
            "1 https://www.ncbi.nlm.nih.gov/nucleotide/NC_051932.1\n",
            "2 https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1\n",
            "3 https://www.ncbi.nlm.nih.gov/nuccore/NC_001665\n",
            "4 https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1\n",
            "5 https://www.ncbi.nlm.nih.gov/nuccore/NC_029849.1\n",
            "6 https://www.ncbi.nlm.nih.gov/nuccore/NC_003081\n",
            "7 https://www.ncbi.nlm.nih.gov/nucleotide/NC_023832.1\n",
            "8 https://www.ncbi.nlm.nih.gov/nuccore/NC_001941\n",
            "9 https://www.ncbi.nlm.nih.gov/nuccore/MZ318036.1\n",
            "10 https://www.ncbi.nlm.nih.gov/nucleotide/NC_024511.2\n",
            "11 https://www.ncbi.nlm.nih.gov/nuccore/MN356385.1\n",
            "12 https://www.ncbi.nlm.nih.gov/nucleotide/NC_022817.1\n",
            "13 https://www.ncbi.nlm.nih.gov/nucleotide/NC_026461.1\n",
            "14 https://www.ncbi.nlm.nih.gov/nuccore/NC_005089\n",
            "15 https://www.ncbi.nlm.nih.gov/nuccore/NC_002784\n",
            "16 https://www.ncbi.nlm.nih.gov/nuccore/NC_009684\n",
            "17 https://www.ncbi.nlm.nih.gov/nuccore/NC_026542.1\n",
            "18 https://www.ncbi.nlm.nih.gov/nuccore/NC_002785\n",
            "19 https://www.ncbi.nlm.nih.gov/nuccore/NC_045042.1\n",
            "20 https://www.ncbi.nlm.nih.gov/nucleotide/NC_027095.1\n",
            "21 https://www.ncbi.nlm.nih.gov/nucleotide/NC_007897.1\n",
            "22 https://www.ncbi.nlm.nih.gov/nucleotide/NC_027938.1\n",
            "23 https://www.ncbi.nlm.nih.gov/nuccore/NC_028302.1\n",
            "24 https://www.ncbi.nlm.nih.gov/nuccore/PP153372.1\n",
            "25 https://www.ncbi.nlm.nih.gov/nuccore/AB080276.1\n",
            "26 https://www.ncbi.nlm.nih.gov/nuccore/PP070534.1\n",
            "27 https://www.ncbi.nlm.nih.gov/nuccore/CM068470.1\n",
            "28 https://www.ncbi.nlm.nih.gov/nuccore/KY986857.1\n",
            "29 https://www.ncbi.nlm.nih.gov/nuccore/OR210205.1\n",
            "30 https://www.ncbi.nlm.nih.gov/nuccore/KY986806.1\n",
            "31 https://www.ncbi.nlm.nih.gov/nuccore/OR917921.1\n",
            "32 https://www.ncbi.nlm.nih.gov/nuccore/OR344290.1\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Y7VcTgsJBXOg"
      },
      "source": [
        "### Step 3: Sequence Download and Preparation\n",
        "\n",
        "The next step in our project involves downloading the mitochondrial DNA sequences for each species and preparing them for analysis.\n",
        "\n",
        "#### Instructions:\n",
        "\n",
        "- **Download mtDNA Sequences**: Write a script to download the mtDNA sequences from the links provided in your dataset. The sequences should be in FASTA format, which is the standard for nucleotide sequences.\n",
        "\n",
        "- **Sequence Labeling**: Properly label each sequence within the FASTA file. This header, starting with '>', should include the species name and any other relevant information (e.g., `>Eudynamys_scolopaceus_NC_060520`). This is crucial for identifying the sequences in subsequent analysis.\n",
        "\n",
        "- **Concatenate Sequences**:\n",
        "    - Create a script to concatenate all downloaded sequences into a single `.fasta` or `.fna` file.\n",
        "    - Ensure each sequence in the file is clearly separated by its header line, which is important for differentiating the sequences of various species.\n",
        "\n",
        "#### Tips for Writing the Download and Concatenation Script:\n",
        "- Use Python libraries such as `httpx` or `requests`, or any other tool you prefer for downloading sequences. For processing FASTA files you can use a wide range of tools. One recommended option is `Biopython` library.\n",
        "- Use a loop to go through each link in the dataset, download the sequence, and append it to your concatenated file.\n",
        "- Maintain the format integrity of the FASTA file, ensuring each sequence is correctly associated with its header.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "id": "nGnlR61LBXOg",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "caea2e2f-5408-401e-de23-0cc6a14e24df"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "33it [00:18,  1.75it/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Sequences downloaded and concatenated successfully!\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        }
      ],
      "source": [
        "from Bio import Entrez, SeqIO\n",
        "\n",
        "# Set your email address for NCBI Entrez\n",
        "Entrez.email = 'mahdi139460@gmail.com'\n",
        "Entrez.api_key = '833b42e2dce5b722fa613375031f546b3e08'\n",
        "\n",
        "# TODO: Write a function to download mtDNA sequences\n",
        "mtDNA_dict = {}\n",
        "\n",
        "def download_mtDNA(link, label):\n",
        "  # Extract the accession number from the link\n",
        "  accession_number = link.split('/')[-1]\n",
        "  # Fetch the sequence from NCBI\n",
        "  with Entrez.efetch(db=\"nuccore\", id=accession_number, rettype=\"fasta\", retmode=\"text\") as handle:\n",
        "      sequence_record = SeqIO.read(handle, \"fasta\")\n",
        "\n",
        "  mtDNA_dict[label] = sequence_record.seq\n",
        "\n",
        "# TODO: Loop through the dataset and download each mtDNA sequence\n",
        "for index, row in tqdm(expanded_dataset.iterrows()):\n",
        "    # Call the download function for each species\n",
        "    label = f'>{\"_\".join(row[\"specie\"].split())}_{row[\"accession_number\"]}'\n",
        "    link = row['mtDNA']\n",
        "\n",
        "    download_mtDNA(link, label)\n",
        "\n",
        "\n",
        "with open(\"concatenated_sequences.fasta\", \"a\") as fasta_file:\n",
        "    # Iterate through each link, download sequence, and write to FASTA file\n",
        "    for label, seq in mtDNA_dict.items():\n",
        "      sequence_record = f'{label}\\n{seq}\\n'\n",
        "      fasta_file.write(sequence_record)\n",
        "\n",
        "print(\"Sequences downloaded and concatenated successfully!\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Mh642ipaBXOh"
      },
      "source": [
        "### Step 4: Sequence Alignment\n",
        "\n",
        "After downloading the mitochondrial DNA sequences, the next critical step is their alignment. This process allows us to compare the sequences and discern the evolutionary relationships among the species.\n",
        "\n",
        "#### Instructions:\n",
        "\n",
        "- **Select an Alignment Tool**: Choose one of the following alignment tools based on your project needs. Each tool has its strengths and is widely used in bioinformatics for multiple sequence alignment.\n",
        "\n",
        "1. **MAFFT**:\n",
        "    - **Brief Introduction**: MAFFT (Multiple Alignment using Fast Fourier Transform) is renowned for its speed and efficiency, particularly suitable for large datasets.\n",
        "    - **Resources**:\n",
        "        - [MAFFT Official Documentation](https://mafft.cbrc.jp/alignment/software/)\n",
        "        - [Example Usage on GitHub](https://github.com/MountainMan12/SARS-Cov2-phylo)\n",
        "        - [Relevant Notebook](https://colab.research.google.com/github/pb3lab/ibm3202/blob/master/tutorials/lab03_phylo.ipynb)\n",
        "\n",
        "2. **Clustal Omega**:\n",
        "    - **Brief Introduction**: Clustal Omega offers high-quality alignments and is user-friendly, ideal for those new to sequence alignment.\n",
        "    - **Resources**:\n",
        "        - [A Python wrapper around Clustal Omega](https://github.com/benchling/clustalo-python)\n",
        "        - [Clustal Omega Official Website](http://www.clustal.org/omega/)\n",
        "\n",
        "3. **MUSCLE**:\n",
        "    - **Brief Introduction**: MUSCLE (Multiple Sequence Comparison by Log-Expectation) is known for its balance between speed and accuracy, making it a versatile choice for various datasets.\n",
        "    - **Resources**:\n",
        "        - [MUSCLE Documentation](https://drive5.com/muscle5/manual/)\n",
        "\n",
        "- **Perform Sequence Alignment**: Utilize your chosen tool to align the downloaded mtDNA sequences. This alignment is foundational for the accurate construction of phylogenetic trees.\n",
        "\n",
        "- **Save Aligned Sequences**: After alignment, save the output in an appropriate format for further analysis in the subsequent steps of this project."
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# from Bio import AlignIO\n",
        "# align = AlignIO.read(\"aligned.fasta\", \"fasta\")"
      ],
      "metadata": {
        "id": "XPb1i36pMoJv"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get install -qq -y mafft"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "dC76fhA68OBb",
        "outputId": "acd5cee4-c7ab-4b7a-fddb-96fa91b629b8"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Extracting templates from packages: 100%\n",
            "Selecting previously unselected package fonts-lato.\n",
            "(Reading database ... 121730 files and directories currently installed.)\n",
            "Preparing to unpack .../00-fonts-lato_2.0-2.1_all.deb ...\n",
            "Unpacking fonts-lato (2.0-2.1) ...\n",
            "Selecting previously unselected package netbase.\n",
            "Preparing to unpack .../01-netbase_6.3_all.deb ...\n",
            "Unpacking netbase (6.3) ...\n",
            "Selecting previously unselected package libclone-perl.\n",
            "Preparing to unpack .../02-libclone-perl_0.45-1build3_amd64.deb ...\n",
            "Unpacking libclone-perl (0.45-1build3) ...\n",
            "Selecting previously unselected package libdata-dump-perl.\n",
            "Preparing to unpack .../03-libdata-dump-perl_1.25-1_all.deb ...\n",
            "Unpacking libdata-dump-perl (1.25-1) ...\n",
            "Selecting previously unselected package libencode-locale-perl.\n",
            "Preparing to unpack .../04-libencode-locale-perl_1.05-1.1_all.deb ...\n",
            "Unpacking libencode-locale-perl (1.05-1.1) ...\n",
            "Selecting previously unselected package libhttp-date-perl.\n",
            "Preparing to unpack .../05-libhttp-date-perl_6.05-1_all.deb ...\n",
            "Unpacking libhttp-date-perl (6.05-1) ...\n",
            "Selecting previously unselected package libfile-listing-perl.\n",
            "Preparing to unpack .../06-libfile-listing-perl_6.14-1_all.deb ...\n",
            "Unpacking libfile-listing-perl (6.14-1) ...\n",
            "Selecting previously unselected package libfont-afm-perl.\n",
            "Preparing to unpack .../07-libfont-afm-perl_1.20-3_all.deb ...\n",
            "Unpacking libfont-afm-perl (1.20-3) ...\n",
            "Selecting previously unselected package libhtml-tagset-perl.\n",
            "Preparing to unpack .../08-libhtml-tagset-perl_3.20-4_all.deb ...\n",
            "Unpacking libhtml-tagset-perl (3.20-4) ...\n",
            "Selecting previously unselected package liburi-perl.\n",
            "Preparing to unpack .../09-liburi-perl_5.10-1_all.deb ...\n",
            "Unpacking liburi-perl (5.10-1) ...\n",
            "Selecting previously unselected package libhtml-parser-perl:amd64.\n",
            "Preparing to unpack .../10-libhtml-parser-perl_3.76-1build2_amd64.deb ...\n",
            "Unpacking libhtml-parser-perl:amd64 (3.76-1build2) ...\n",
            "Selecting previously unselected package libio-html-perl.\n",
            "Preparing to unpack .../11-libio-html-perl_1.004-2_all.deb ...\n",
            "Unpacking libio-html-perl (1.004-2) ...\n",
            "Selecting previously unselected package liblwp-mediatypes-perl.\n",
            "Preparing to unpack .../12-liblwp-mediatypes-perl_6.04-1_all.deb ...\n",
            "Unpacking liblwp-mediatypes-perl (6.04-1) ...\n",
            "Selecting previously unselected package libhttp-message-perl.\n",
            "Preparing to unpack .../13-libhttp-message-perl_6.36-1_all.deb ...\n",
            "Unpacking libhttp-message-perl (6.36-1) ...\n",
            "Selecting previously unselected package libhtml-form-perl.\n",
            "Preparing to unpack .../14-libhtml-form-perl_6.07-1_all.deb ...\n",
            "Unpacking libhtml-form-perl (6.07-1) ...\n",
            "Selecting previously unselected package libhtml-tree-perl.\n",
            "Preparing to unpack .../15-libhtml-tree-perl_5.07-2_all.deb ...\n",
            "Unpacking libhtml-tree-perl (5.07-2) ...\n",
            "Selecting previously unselected package libhtml-format-perl.\n",
            "Preparing to unpack .../16-libhtml-format-perl_2.12-1.1_all.deb ...\n",
            "Unpacking libhtml-format-perl (2.12-1.1) ...\n",
            "Selecting previously unselected package libhttp-cookies-perl.\n",
            "Preparing to unpack .../17-libhttp-cookies-perl_6.10-1_all.deb ...\n",
            "Unpacking libhttp-cookies-perl (6.10-1) ...\n",
            "Selecting previously unselected package libhttp-daemon-perl.\n",
            "Preparing to unpack .../18-libhttp-daemon-perl_6.13-1ubuntu0.1_all.deb ...\n",
            "Unpacking libhttp-daemon-perl (6.13-1ubuntu0.1) ...\n",
            "Selecting previously unselected package libhttp-negotiate-perl.\n",
            "Preparing to unpack .../19-libhttp-negotiate-perl_6.01-1_all.deb ...\n",
            "Unpacking libhttp-negotiate-perl (6.01-1) ...\n",
            "Selecting previously unselected package perl-openssl-defaults:amd64.\n",
            "Preparing to unpack .../20-perl-openssl-defaults_5build2_amd64.deb ...\n",
            "Unpacking perl-openssl-defaults:amd64 (5build2) ...\n",
            "Selecting previously unselected package libnet-ssleay-perl:amd64.\n",
            "Preparing to unpack .../21-libnet-ssleay-perl_1.92-1build2_amd64.deb ...\n",
            "Unpacking libnet-ssleay-perl:amd64 (1.92-1build2) ...\n",
            "Selecting previously unselected package libio-socket-ssl-perl.\n",
            "Preparing to unpack .../22-libio-socket-ssl-perl_2.074-2_all.deb ...\n",
            "Unpacking libio-socket-ssl-perl (2.074-2) ...\n",
            "Selecting previously unselected package libnet-http-perl.\n",
            "Preparing to unpack .../23-libnet-http-perl_6.22-1_all.deb ...\n",
            "Unpacking libnet-http-perl (6.22-1) ...\n",
            "Selecting previously unselected package libtry-tiny-perl.\n",
            "Preparing to unpack .../24-libtry-tiny-perl_0.31-1_all.deb ...\n",
            "Unpacking libtry-tiny-perl (0.31-1) ...\n",
            "Selecting previously unselected package libwww-robotrules-perl.\n",
            "Preparing to unpack .../25-libwww-robotrules-perl_6.02-1_all.deb ...\n",
            "Unpacking libwww-robotrules-perl (6.02-1) ...\n",
            "Selecting previously unselected package libwww-perl.\n",
            "Preparing to unpack .../26-libwww-perl_6.61-1_all.deb ...\n",
            "Unpacking libwww-perl (6.61-1) ...\n",
            "Selecting previously unselected package liblwp-protocol-https-perl.\n",
            "Preparing to unpack .../27-liblwp-protocol-https-perl_6.10-1_all.deb ...\n",
            "Unpacking liblwp-protocol-https-perl (6.10-1) ...\n",
            "Selecting previously unselected package libnet-smtp-ssl-perl.\n",
            "Preparing to unpack .../28-libnet-smtp-ssl-perl_1.04-1_all.deb ...\n",
            "Unpacking libnet-smtp-ssl-perl (1.04-1) ...\n",
            "Selecting previously unselected package libmailtools-perl.\n",
            "Preparing to unpack .../29-libmailtools-perl_2.21-1_all.deb ...\n",
            "Unpacking libmailtools-perl (2.21-1) ...\n",
            "Selecting previously unselected package rubygems-integration.\n",
            "Preparing to unpack .../30-rubygems-integration_1.18_all.deb ...\n",
            "Unpacking rubygems-integration (1.18) ...\n",
            "Selecting previously unselected package ruby3.0.\n",
            "Preparing to unpack .../31-ruby3.0_3.0.2-7ubuntu2.4_amd64.deb ...\n",
            "Unpacking ruby3.0 (3.0.2-7ubuntu2.4) ...\n",
            "Selecting previously unselected package ruby-rubygems.\n",
            "Preparing to unpack .../32-ruby-rubygems_3.3.5-2_all.deb ...\n",
            "Unpacking ruby-rubygems (3.3.5-2) ...\n",
            "Selecting previously unselected package ruby.\n",
            "Preparing to unpack .../33-ruby_1%3a3.0~exp1_amd64.deb ...\n",
            "Unpacking ruby (1:3.0~exp1) ...\n",
            "Selecting previously unselected package rake.\n",
            "Preparing to unpack .../34-rake_13.0.6-2_all.deb ...\n",
            "Unpacking rake (13.0.6-2) ...\n",
            "Selecting previously unselected package ruby-net-telnet.\n",
            "Preparing to unpack .../35-ruby-net-telnet_0.1.1-2_all.deb ...\n",
            "Unpacking ruby-net-telnet (0.1.1-2) ...\n",
            "Selecting previously unselected package ruby-webrick.\n",
            "Preparing to unpack .../36-ruby-webrick_1.7.0-3_all.deb ...\n",
            "Unpacking ruby-webrick (1.7.0-3) ...\n",
            "Selecting previously unselected package ruby-xmlrpc.\n",
            "Preparing to unpack .../37-ruby-xmlrpc_0.3.2-1ubuntu0.1_all.deb ...\n",
            "Unpacking ruby-xmlrpc (0.3.2-1ubuntu0.1) ...\n",
            "Selecting previously unselected package libruby3.0:amd64.\n",
            "Preparing to unpack .../38-libruby3.0_3.0.2-7ubuntu2.4_amd64.deb ...\n",
            "Unpacking libruby3.0:amd64 (3.0.2-7ubuntu2.4) ...\n",
            "Selecting previously unselected package lynx-common.\n",
            "Preparing to unpack .../39-lynx-common_2.9.0dev.10-1_all.deb ...\n",
            "Unpacking lynx-common (2.9.0dev.10-1) ...\n",
            "Selecting previously unselected package mafft.\n",
            "Preparing to unpack .../40-mafft_7.490-1_amd64.deb ...\n",
            "Unpacking mafft (7.490-1) ...\n",
            "Selecting previously unselected package mailcap.\n",
            "Preparing to unpack .../41-mailcap_3.70+nmu1ubuntu1_all.deb ...\n",
            "Unpacking mailcap (3.70+nmu1ubuntu1) ...\n",
            "Selecting previously unselected package mime-support.\n",
            "Preparing to unpack .../42-mime-support_3.66_all.deb ...\n",
            "Unpacking mime-support (3.66) ...\n",
            "Selecting previously unselected package libauthen-sasl-perl.\n",
            "Preparing to unpack .../43-libauthen-sasl-perl_2.1600-1.1_all.deb ...\n",
            "Unpacking libauthen-sasl-perl (2.1600-1.1) ...\n",
            "Selecting previously unselected package lynx.\n",
            "Preparing to unpack .../44-lynx_2.9.0dev.10-1_amd64.deb ...\n",
            "Unpacking lynx (2.9.0dev.10-1) ...\n",
            "Setting up libhttp-date-perl (6.05-1) ...\n",
            "Setting up fonts-lato (2.0-2.1) ...\n",
            "Setting up libfile-listing-perl (6.14-1) ...\n",
            "Setting up libfont-afm-perl (1.20-3) ...\n",
            "Setting up mafft (7.490-1) ...\n",
            "Setting up libclone-perl (0.45-1build3) ...\n",
            "Setting up libhtml-tagset-perl (3.20-4) ...\n",
            "Setting up libauthen-sasl-perl (2.1600-1.1) ...\n",
            "Setting up liblwp-mediatypes-perl (6.04-1) ...\n",
            "Setting up libtry-tiny-perl (0.31-1) ...\n",
            "Setting up perl-openssl-defaults:amd64 (5build2) ...\n",
            "Setting up libencode-locale-perl (1.05-1.1) ...\n",
            "Setting up rubygems-integration (1.18) ...\n",
            "Setting up libdata-dump-perl (1.25-1) ...\n",
            "Setting up ruby-net-telnet (0.1.1-2) ...\n",
            "Setting up libio-html-perl (1.004-2) ...\n",
            "Setting up lynx-common (2.9.0dev.10-1) ...\n",
            "Setting up ruby-webrick (1.7.0-3) ...\n",
            "Setting up netbase (6.3) ...\n",
            "Setting up lynx (2.9.0dev.10-1) ...\n",
            "update-alternatives: using /usr/bin/lynx to provide /usr/bin/www-browser (www-browser) in auto mode\n",
            "Setting up mailcap (3.70+nmu1ubuntu1) ...\n",
            "Setting up ruby-xmlrpc (0.3.2-1ubuntu0.1) ...\n",
            "Setting up liburi-perl (5.10-1) ...\n",
            "Setting up libhttp-message-perl (6.36-1) ...\n",
            "Setting up libnet-ssleay-perl:amd64 (1.92-1build2) ...\n",
            "Setting up libhttp-negotiate-perl (6.01-1) ...\n",
            "Setting up mime-support (3.66) ...\n",
            "Setting up libhttp-cookies-perl (6.10-1) ...\n",
            "Setting up libnet-http-perl (6.22-1) ...\n",
            "Setting up libwww-robotrules-perl (6.02-1) ...\n",
            "Setting up libhttp-daemon-perl (6.13-1ubuntu0.1) ...\n",
            "Setting up libhtml-parser-perl:amd64 (3.76-1build2) ...\n",
            "Setting up libio-socket-ssl-perl (2.074-2) ...\n",
            "Setting up libhtml-form-perl (6.07-1) ...\n",
            "Setting up libhtml-tree-perl (5.07-2) ...\n",
            "Setting up libhtml-format-perl (2.12-1.1) ...\n",
            "Setting up libnet-smtp-ssl-perl (1.04-1) ...\n",
            "Setting up libmailtools-perl (2.21-1) ...\n",
            "Setting up libwww-perl (6.61-1) ...\n",
            "Setting up ruby3.0 (3.0.2-7ubuntu2.4) ...\n",
            "Setting up liblwp-protocol-https-perl (6.10-1) ...\n",
            "Setting up ruby (1:3.0~exp1) ...\n",
            "Setting up rake (13.0.6-2) ...\n",
            "Setting up libruby3.0:amd64 (3.0.2-7ubuntu2.4) ...\n",
            "Setting up ruby-rubygems (3.3.5-2) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for fontconfig (2.13.1-4.2ubuntu5) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "id": "oqubMjy9BXOh",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "d4fd185d-2cda-4461-b123-645c4ec7a15d"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/Bio/Application/__init__.py:40: BiopythonDeprecationWarning: The Bio.Application modules and modules relying on it have been deprecated.\n",
            "\n",
            "Due to the on going maintenance burden of keeping command line application\n",
            "wrappers up to date, we have decided to deprecate and eventually remove these\n",
            "modules.\n",
            "\n",
            "We instead now recommend building your command line and invoking it directly\n",
            "with the subprocess module.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "mafft concatenated_sequences.fasta\n"
          ]
        }
      ],
      "source": [
        "# TODO: Install and import the alignment tool\n",
        "# !pip install mafft\n",
        "from Bio.Align.Applications import MafftCommandline\n",
        "\n",
        "# TODO: Perform the sequence alignment\n",
        "def perform_alignment(input_file, output_file):\n",
        "    # Code to align sequences using the chosen tool\n",
        "    mafft_cline = MafftCommandline(input=input_file)\n",
        "    print(mafft_cline)\n",
        "    stdout, stderr = mafft_cline()\n",
        "    with open(output_file, \"w\") as handle:\n",
        "      handle.write(stdout)\n",
        "\n",
        "# TODO: Align your downloaded sequences\n",
        "perform_alignment('concatenated_sequences.fasta', 'aligned_sequences.fasta')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "asgJa3HnBXOh"
      },
      "source": [
        "### Step 5: Phylogenetic Tree Construction\n",
        "\n",
        "The next phase in our project involves constructing phylogenetic trees to visualize and analyze the evolutionary relationships among the species. We will use three distinct methods, each providing unique insights.\n",
        "\n",
        "#### Phylogenetic Tree Construction Methods:\n",
        "\n",
        "1. **Bayesian Inference Trees**:\n",
        "    - **Overview**: This method uses Bayesian statistics to estimate the likelihood of different evolutionary histories. It's particularly useful for its ability to estimate branch lengths and support values.\n",
        "    - **Tools**: MrBayes, BEAST\n",
        "        - MrBayes ([Official Website](https://nbisweden.github.io/MrBayes/manual.html/)) is widely recognized for its robustness in Bayesian inference.\n",
        "        - BEAST2 ([BEAST Software](https://www.beast2.org/)) is another powerful tool, offering advanced features for complex evolutionary models.\n",
        "\n",
        "2. **Maximum Likelihood Trees**:\n",
        "    - **Overview**: Maximum Likelihood methods evaluate tree topologies based on the likelihood of observed data given a tree model. It's known for its statistical rigor and accuracy.\n",
        "    - **Tools**: RAxML, PhyML\n",
        "        - RAxML ([RAxML GitHub](https://github.com/stamatak/standard-RAxML)) is preferred for large datasets due to its efficiency.\n",
        "        - PhyML ([PhyML Documentation](http://www.atgc-montpellier.fr/phyml/)) offers a balance of speed and accuracy, with a user-friendly interface.\n",
        "\n",
        "3. **Neighbor-Joining Trees**:\n",
        "    - **Overview**: The Neighbor-Joining method is a distance-based approach that constructs phylogenetic trees by evaluating the genetic distance between sequences. It is known for its speed and simplicity, making it well-suited for initial exploratory analyses.\n",
        "    - **Tools**:\n",
        "        - MEGA: A versatile tool specifically used here for constructing Neighbor-Joining trees. It's recognized for its ease of use and effectiveness in phylogenetic analysis. [MEGA Software](https://www.megasoftware.net/)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "id": "vOQqNFDnBXOh",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 367
        },
        "outputId": "8362b4f5-2fa3-401a-f7ad-e3ac130c3721"
      },
      "outputs": [
        {
          "output_type": "error",
          "ename": "TypeError",
          "evalue": "DistanceTreeConstructor.build_tree() got an unexpected keyword argument 'distance_matrix'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-12-611cade4236c>\u001b[0m in \u001b[0;36m<cell line: 69>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     67\u001b[0m \u001b[0;31m# Call the construct_tree_NJ function with the alignment object\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     68\u001b[0m \u001b[0mtree_NJ\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconstruct_tree_NJ\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malignment\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 69\u001b[0;31m \u001b[0mtree_ML\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconstruct_tree_ML\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malignment\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     70\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     71\u001b[0m \u001b[0;31m# Visualize the resulting tree\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m<ipython-input-12-611cade4236c>\u001b[0m in \u001b[0;36mconstruct_tree_ML\u001b[0;34m(aligned_sequences)\u001b[0m\n\u001b[1;32m     27\u001b[0m     \u001b[0;31m# Construct the tree using Maximum Likelihood method\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m     \u001b[0mconstructor\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mDistanceTreeConstructor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 29\u001b[0;31m     \u001b[0mtree_ML\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconstructor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuild_tree\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maligned_sequences\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdistance_matrix\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdistance_matrix\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     30\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     31\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mtree_ML\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mTypeError\u001b[0m: DistanceTreeConstructor.build_tree() got an unexpected keyword argument 'distance_matrix'"
          ]
        }
      ],
      "source": [
        "# TODO: Import necessary libraries for tree construction\n",
        "from Bio import Phylo, SeqIO\n",
        "from Bio.Align import MultipleSeqAlignment\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "from Bio.Seq import Seq\n",
        "import subprocess\n",
        "# import matplotlib.pyplot as plt\n",
        "from Bio.Phylo.TreeConstruction import DistanceCalculator, DistanceTreeConstructor\n",
        "\n",
        "# TODO: Construct a phylogenetic tree using the Neighbor Joining method\n",
        "def construct_tree_NJ(aligned_sequences):\n",
        "    # Implement the tree construction using Neighbor Joining\n",
        "    calculator = DistanceCalculator('identity')\n",
        "    dm = calculator.get_distance(aligned_sequences)\n",
        "    constructor = DistanceTreeConstructor()\n",
        "    tree = constructor.nj(dm)\n",
        "    return tree\n",
        "\n",
        "# TODO: Repeat the process for Maximum Likelihood and Supertree methods\n",
        "# def construct_tree_ML(aligned_sequences):\n",
        "    # Implement the tree construction using Maximum Likelihood\n",
        "from Bio.Phylo.TreeConstruction import DistanceCalculator\n",
        "from Bio import AlignIO\n",
        "from Bio.Phylo.Phyml import PhymlCommandline\n",
        "from Phylo.PhyloXML import PhymlCommandline\n",
        "# Define a function to construct a phylogenetic tree using Maximum Likelihood\n",
        "def construct_tree_ML(aligned_sequences):\n",
        "    # Calculate the distance matrix\n",
        "    calculator = DistanceCalculator('identity')\n",
        "    distance_matrix = calculator.get_distance(aligned_sequences)\n",
        "\n",
        "    # Save the distance matrix to a temporary file\n",
        "    temp_dist_file = \"distance_matrix.txt\"\n",
        "    calculator.write_to_file(distance_matrix, temp_dist_file)\n",
        "\n",
        "    # Construct the tree using Phyml\n",
        "    phyml_cline = PhymlCommandline(input=temp_dist_file)\n",
        "    phyml_cline.outfile = \"tree_ML.phy\"\n",
        "    phyml_cline.run()\n",
        "\n",
        "    # Read the resulting ML tree\n",
        "    tree_ML = Phylo.read(\"tree_ML.phy_phyml_tree.txt\", \"newick\")\n",
        "\n",
        "    return tree_ML\n",
        "\n",
        "# Example usage:\n",
        "# aligned_sequences = AlignIO.read(\"aligned_sequences.fasta\", \"fasta\")\n",
        "# tree_ML = construct_tree_ML(aligned_sequences)\n",
        "\n",
        "\n",
        "# def construct_tree_BI(gene_trees):\n",
        "    # Implement the construction of a tree using bayes inference\n",
        "\n",
        "\n",
        "# Define a function to construct a phylogenetic tree using Bayesian inference with MrBayes\n",
        "def construct_tree_BI(aligned_sequences):\n",
        "    # Write the aligned sequences to a temporary file in PHYLIP format\n",
        "    temp_file = \"aligned_sequences.phy\"\n",
        "    AlignIO.write(aligned_sequences, temp_file, \"phylip\")\n",
        "\n",
        "    # Define the MrBayes command\n",
        "    mrbayes_command = f\"mb {temp_file}\"\n",
        "\n",
        "    # Execute the MrBayes command using subprocess\n",
        "    subprocess.run(mrbayes_command, shell=True)\n",
        "\n",
        "# TODO: Visualize and save the constructed trees\n",
        "# Phylo.draw(tree, do_show=False)\n",
        "# Phylo.write(tree, 'tree_output.xml', 'phyloxml')\n",
        "\n",
        "\n",
        "# Parse the FASTA file and extract aligned sequences\n",
        "aligned_sequences = []\n",
        "fasta_file = \"aligned_sequences.fasta\"\n",
        "\n",
        "for i, record in enumerate(SeqIO.parse(fasta_file, \"fasta\")):\n",
        "    # Append a unique index to the ID to ensure uniqueness\n",
        "    record = SeqRecord(Seq(record.seq), id=record.id, description=\"\")\n",
        "    aligned_sequences.append(record)\n",
        "\n",
        "\n",
        "# Convert the list of SeqRecord objects to a MultipleSeqAlignment object\n",
        "alignment = MultipleSeqAlignment(aligned_sequences)\n",
        "\n",
        "# Call the construct_tree_NJ function with the alignment object\n",
        "tree_NJ = construct_tree_NJ(alignment)\n",
        "tree_ML = construct_tree_ML(alignment)\n",
        "\n",
        "# Visualize the resulting tree\n",
        "Phylo.draw(tree_NJ)\n",
        "Phylo.draw(tree_ML)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "chhARWC2BXOi"
      },
      "source": [
        "### Step 6: In-Depth Phylogenetic Tree Visualization\n",
        "\n",
        "Having constructed phylogenetic trees using different methods, our next task is to visualize these trees effectively. This step is crucial for interpreting the results and communicating our findings.\n",
        "\n",
        "#### Visualization Tools:\n",
        "\n",
        "1. **FigTree**:\n",
        "    - **Overview**: FigTree is designed for the graphical representation of phylogenetic trees. It's excellent for creating publication-ready visualizations.\n",
        "    - **Resource**: [FigTree Tool](http://tree.bio.ed.ac.uk/software/figtree/)\n",
        "    - **Usage**: Use FigTree to add detailed annotations, adjust branch colors, and format tree layouts for clear, interpretable visualizations.\n",
        "\n",
        "2. **iTOL (Interactive Tree Of Life)**:\n",
        "    - **Overview**: iTOL is a web-based tool for the display, annotation, and management of phylogenetic trees, offering extensive customization options.\n",
        "    - **Resource**: [iTOL Website](https://itol.embl.de/)\n",
        "    - **Usage**: Ideal for interactive tree visualizations. It allows users to explore different layers of data through their tree, such as adding charts or color-coding branches.\n",
        "\n",
        "3. **Dendroscope**:\n",
        "    - **Overview**: Dendroscope is a software program for viewing and editing phylogenetic trees, particularly useful for large datasets.\n",
        "    - **Resource**: [Dendroscope Download](https://uni-tuebingen.de/fakultaeten/mathematisch-naturwissenschaftliche-fakultaet/fachbereiche/informatik/lehrstuehle/algorithms-in-bioinformatics/software/dendroscope/)\n",
        "    - **Usage**: Utilize Dendroscope when dealing with large and complex trees or when you need to compare multiple trees side-by-side.\n",
        "\n",
        "#### Task:\n",
        "\n",
        "- **Visualize Each Tree**: Use one or more of the above tools to visualize the phylogenetic trees you constructed using Bayesian inference, maximum likelihood, and neighbor-joining methods.\n",
        "- **Highlight Differences**: Focus on highlighting the differences and similarities between the trees obtained from the different methods. Pay attention to tree topology, branch lengths, and any notable patterns.\n",
        "- **Interpretation and Presentation**: Aim for visualizations that are not only accurate but also interpretable and visually appealing. This will enhance the clarity of your work."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "8XJ9SPizBXOi"
      },
      "outputs": [],
      "source": [
        "### TODO ####\n",
        "# You code for visualization of each tree\n",
        "### TODO ###"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ysa4xKNOBXOi"
      },
      "source": [
        "### Cross-Disciplinary Applications (Optional)\n",
        "\n",
        "This is an optional part with bonus, relative to the depth of your analysis. Refer to the first part of this notebook. You have complete freedom to do this part anyway you like, but to gain a portion of the bonus score for this section, a bare minimum effort is required."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Vj1o2KYYBXOi"
      },
      "source": [
        "### Conclusion and Reflective Insights\n",
        "\n",
        "As we conclude our exploration of phylogenetic tree construction and analysis, let's reflect on the insights learned from this task and consider questions that emerge from our findings.\n",
        "\n",
        "#### Interpretation of Results:\n",
        "\n",
        "- Reflect on the phylogenetic trees produced by each method (Bayesian inference, maximum likelihood, and neighbor-joining). Consider how the differences in tree topology might offer varied perspectives on the evolutionary relationships among the species.\n",
        "\n",
        "#### Questions to Ponder:\n",
        "\n",
        "1. **Species Divergence**: Based on the trees, which species appear to have the most ancient divergence? How might this information contribute to our understanding of their evolutionary history?\n",
        "   \n",
        "2. **Common Ancestors**: Are there any unexpected pairings or groupings of species that suggest a closer evolutionary relationship than previously thought? How could this reshape our understanding of these species' evolutionary paths?\n",
        "\n",
        "3. **Methodology Insights**: Considering the discrepancies between the trees generated by different methods, what might this tell us about the limitations and strengths of each phylogenetic analysis method?\n",
        "\n",
        "4. **Conservation Implications**: Considering the evolutionary relationships revealed in your phylogenetic analysis, what insights can be gained for conservation strategies? Specifically, how could understanding the close evolutionary ties between species, which might be facing distinct environmental challenges, guide targeted conservation efforts?\n",
        "\n",
        "<blockquote style=\"font-family:Arial; color:red; font-size:16px; border-left:0px solid red; padding: 10px;\">\n",
        "    <strong>Don't forget to answer these questions!</strong>\n",
        "</blockquote>\n"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.7"
    },
    "colab": {
      "provenance": []
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}